### PR TITLE
fix: run dependent upstream blocks when you run a single block

### PR DESCRIPTION
**Pull request by jamesbhobbs**
NB consider whether to make this optional

## Summary by CodeRabbit

* **New Features**
  * Running blocks can include resolved upstream dependencies; CLI dry-run shows upstream ordering.
  * Support running multiple target block IDs together; explicit list takes precedence over single-block target.
  * Notebook-scoped upstream analysis (when notebook not specified, scope is confined).

* **Bug Fixes / UX**
  * Clearer errors when blocks are missing or non-executable; graceful handling of partial/missing upstream deps.

* **Tests**
  * Added tests for upstream resolution, multi-block execution, precedence, filtering, and edge cases.

### CODE REVIEW

File: packages/reactivity/src/dag-analyzer.ts [hunk 1]
```diff
@@ -39,3 +39,43 @@ function getDownstreamBlocks(dag: BlockDependencyDag, blocksIds: string[], visit
...
+    nodes: dag.nodes.filter(node => node.order <= maxOrder),
+  }
+
+  return getUpstreamBlocks(filteredDag, blocksIds)
+}
+
+function getUpstreamBlocks(dag: BlockDependencyDag, blocksIds: string[], visited: Set<string> = new Set()): string[] {
+  const inputVariables = dag.nodes.filter(node => blocksIds.includes(node.id)).flatMap(node => node.inputVariables)
+
+  const blocksThatDefineVariables = dag.nodes
```
Comment (Artmann) [inline packages/reactivity/src/dag-analyzer.ts:66]: How does this work? Will this only find blocks where a variable is defined? Don't we need to walk up the three and grab all the blocks along the way?

For examle if we have three blocks:

```
A
df = pd.Dataframe()

B
df["a"] = 42

C 
print(df)
```

If we run `C`, we would like to execute `A -> B -> C` to get the correct result.

Comment (jamesbhobbs) [reply packages/reactivity/src/dag-analyzer.ts:66]: They are all found by DAG and C has dependencies A+B. Execution order is currently based on order in the notebook and thus it will run A->B->C.
